### PR TITLE
ref(perf_issues): Initial post_process_group refactor (ISP-11)

### PR DESCRIPTION
**Pull request by rjo100**
Initial refactor of `post_process_group`. Will refactor `group_inbox` stuff in a subsequent PR. Then, once we have confidence in the changes, will upgrade to a more robust pipeline system

### CODE REVIEW

File: src/sentry/tasks/post_process.py [hunk 1]
```diff
@@ -564,6 +445,160 @@ def process_snoozes(group):
     return False
 
 
+def process_event(cache_key, data, group_id):
```
Comment (wedamija) [inline src/sentry/tasks/post_process.py:448]: Looks like `cache_key` is unused

File: src/sentry/tasks/post_process.py [hunk 2]
```diff
@@ -564,6 +445,160 @@ def process_snoozes(group):
...
+        # TODO(dcramer): ideally this would fanout, but serializing giant
+        # objects back and forth isn't super efficient
+        for callback, futures in rp.apply():
+            has_alert = True
+            safe_execute(callback, event, futures, _with_transaction=False)
+
+    return has_alert
+
+
+def process_commits(event):
```
Comment (wedamija) [inline src/sentry/tasks/post_process.py:502]: We never call this

**Review by wedamija — COMMENTED** [review]
Generally looking good. Scary that our tests didn't catch the `process_commits` thing

File: src/sentry/tasks/post_process.py (hunk 1 above)
Comment (rjo100) [reply src/sentry/tasks/post_process.py:448]: good catch that was a regression from an earlier fix

File: src/sentry/tasks/post_process.py (hunk 2 above)
Comment (rjo100) [reply src/sentry/tasks/post_process.py:502]: good catch x2. adding back. linter should have caught it too

File: src/sentry/tasks/post_process.py [hunk 3]
```diff
@@ -564,6 +446,160 @@ def process_snoozes(group):
     return False
 
 
+def process_event(data, group_id):
```
Comment (barkbarkimashark) [inline src/sentry/tasks/post_process.py:449]: Would it make sense to add typing to the newly created methods?

File: src/sentry/tasks/post_process.py [hunk 4]
```diff
@@ -564,6 +446,160 @@ def process_snoozes(group):
...
+    event.group_id = event.group.id
+
+    # We fetch buffered updates to group aggregates here and populate them on the Group. This
+    # helps us avoid problems with processing group ignores and alert rules that rely on these
+    # stats.
+    with sentry_sdk.start_span(op="tasks.post_process_group.fetch_buffered_group_stats"):
+        fetch_buffered_group_stats(event.group)
+
+    event.group.project = event.project
+    event.group.project.set_cached_field_value("organization", event.project.organization)
```
Comment (barkbarkimashark) [inline src/sentry/tasks/post_process.py:484]: Would it make sense to differentiate the `process_*` methods into two types:
* mutating input data (`update_*`)
* non-mutating inputs (`process_*`)

I'm always weary of methods that mutates the inputs and I'd be nice to have that explicitly called out when scanning through the code.

File: src/sentry/tasks/post_process.py (hunk 3 above)
Comment (rjo100) [reply src/sentry/tasks/post_process.py:449]: sure thing

File: src/sentry/tasks/post_process.py (hunk 4 above)
Comment (rjo100) [reply src/sentry/tasks/post_process.py:484]: makes sense. will modify

**Review by wedamija — APPROVED** [review]
This looks good to me, other than the various small tidy ups from other comments.

When you merge this, make sure you're around to monitor the deploy, just in case we cause any issues. Probably ping me on slack so I can keep an eye out too

**Pull request merged**